rpcs3/rpcs3qt/infinity_dialog.cpp

```diff
@@ -0,0 +1,821 @@
+			if (file_path.isEmpty())
+			{
+				return;
+			}
+			if (!create_blank_figure(fig_num, series))
+			{
+				QMessageBox::warning(this, tr("Failed to create figure file!"), tr("Failed to create figure file:\n%1").arg(file_path), QMessageBox::Ok);
+				return;
+			}
+
+			last_figure_path = QFileInfo(file_path).absolutePath() + "/";
+			accept();
+		});
+
+	connect(btn_cancel, &QAbstractButton::clicked, this, &QDialog::reject);
+
+	connect(co_compl, QOverload<const QString&>::of(&QCompleter::activated), [=](const QString& text)
+		{
+			combo_figlist->setCurrentText(text);
+			combo_figlist->setCurrentIndex(combo_figlist->findText(text));
+		});
+}
+
+bool figure_creator_dialog::create_blank_figure(u32 character, u8 series)
+{
+	infinity_log.notice("File path: %s Character: %d Series: %d", file_path.toStdString(), character, series);
+	fs::file inf_file(file_path.toStdString(), fs::read + fs::write + fs::create);
+	if (!inf_file)
+	{
+		return false;
+	}
+	// Create a 320 byte file with standard NFC read/write permissions
+	std::array<u8, 0x14 * 0x10> file_data{};
+	u32 first_block = 0x17878E;
+	u32 other_blocks = 0x778788;
+	for (s8 i = 2; i >= 0; i--)
+	{
+		file_data[0x38 - i] = u8((first_block >> i * 8) & 0xFF);
+	}
+	for (u32 index = 1; index < 0x05; index++)
+	{
+		for (s8 i = 2; i >= 0; i--)
+		{
+			file_data[((index * 0x40) + 0x38) - i] = u8((other_blocks >> i * 8) & 0xFF);
+		}
+	}
+	// Create the vector to calculate the SHA1 hash with
+	std::vector<u8> sha1_calc = {SHA1_CONSTANT.begin(), SHA1_CONSTANT.end() - 1};
+	// Generate random UID, used for AES encrypt/decrypt
+	std::array<u8, 16> uid_data = {0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x89, 0x44, 0x00, 0xC2};
+	for (s8 i = 0; i < 7; i++)
+	{
+		sha1_calc.push_back(uid_data[i]);
+	}
+
+	std::array<u8, 16> figure_data = {0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+		0x00, 0x00, 0x00, series, 0xD1, 0x1F};
+
+	// Figure Number, input by end user
+	figure_data[1] = u8((character >> 16) & 0xFF);
+	figure_data[2] = u8((character >> 8) & 0xFF);
+	figure_data[3] = u8(character & 0xFF);
+
+	// Manufacture date, formatted as YY/MM/DD. Set to release date of figure's series
+	if (series == 1)
+	{
+		figure_data[4] = 0x0D;
+		figure_data[5] = 0x08;
+		figure_data[6] = 0x12;
+	}
+	else if (series == 2)
+	{
+		figure_data[4] = 0x0E;
+		figure_data[5] = 0x09;
+		figure_data[6] = 0x12;
+	}
+	else if (series == 3)
+	{
+		figure_data[4] = 0x0F;
+		figure_data[5] = 0x08;
+		figure_data[6] = 0x1C;
+	}
+
+	u32 checksum = infinity_crc32(0, figure_data.data(), 12);
+	for (s8 i = 3; i >= 0; i--)
+	{
+		figure_data[15 - i] = u8((checksum >> i * 8) & 0xFF);
+	}
+
+	if (figure_data[1] == 0x00)
+		return false;
+
+	sha1_context ctx;
+	u8 output[20];
+
+	sha1_starts(&ctx);
+	sha1_update(&ctx, sha1_calc.data(), sha1_calc.size());
+	sha1_finish(&ctx, output);
+
+	u8 key[0x10];
+	for (int i = 0; i < 4; i++)
+	{
+		for (int x = 3; x >= 0; x--)
```

Comment:
why is this in reverse?

Reply:
Just a quirk with Disney Infinity - their data is encrypted by an AES key which is the first 16 bytes of an SHA1 digest - but every 4 bytes needs to be reversed, will add a comment in here explaining this

Reply:
I don't see how how this affects how you set the bytes.
a regular forward loop should be identical unless you set the same bytes twice

Reply:
Sorry, have been confusing myself - updating the reversed ones now